### PR TITLE
feat(cli): Add hyperlane warp extend

### PR DESCRIPTION
**Pull request by ltyu**
### Description
- Adds "extend warp route" logic to `hyperlane warp update`
- To extend, provide the existing `warp-route-deployment.yaml` with the additional `TokenRouterConfig`

For example, considering `hyperlane warp deploy` is called with this config:
```yaml
alpha:
  interchainSecurityModule:
    modules:
      - relayer: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: trustedRelayerIsm
      - domains: {}
        owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: '0xe04e5D673F3f253BF43A82ce66Bba0Ae771CF09B'
  owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
  type: native
```

To extend, add a new TokenRouter config:
```yaml
alpha:
  interchainSecurityModule:
    modules:
      - relayer: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: trustedRelayerIsm
      - domains: {}
        owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: '0xe04e5D673F3f253BF43A82ce66Bba0Ae771CF09B'
  owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
  type: native
arbitrum:
  interchainSecurityModule:
    modules:
      - relayer: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: trustedRelayerIsm
      - domains: {}
        owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: '0x979Ca5202784112f4738403dBec5D0F3B9daabB9'
  owner: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'
  type: native
```

The newly deployed `arbitrum` `TokenRouterConfig` will deploy the Router, and then enroll the existing `alpha` Router with each other.

### Drive by
- Fix native deriving to send 0 wei instead of 1

### Related issues

https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/4078

### Testing
Manual 
- Tested with deploying 
  - native and extending with native
  - native and extending with native, and native
  - native and extending with synthetic
- use `cast abi-decode "domain()(uint32[])" $(cast call $TOKEN_ROUTER  "domains()")` to test enrollment

E2e test

### CODE REVIEW

File: typescript/cli/src/deploy/warp.ts [hunk 1]
```diff
@@ -154,7 +173,7 @@ async function executeDeploy(params: DeployParams, apiKeys: ChainMap<string>) {
 
   const deployer = warpDeployConfig.isNft
     ? new HypERC721Deployer(multiProvider)
-    : new HypERC20Deployer(multiProvider);
+    : new HypERC20Deployer(multiProvider); // @TODO replace with EvmERC20WarpModule
```
Comment (nbayindirli) [inline typescript/cli/src/deploy/warp.ts:176]: nit
```suggestion
    : new HypERC20Deployer(multiProvider); // TODO: replace with EvmERC20WarpModule
```

File: typescript/cli/src/deploy/warp.ts [hunk 2]
```diff
@@ -387,48 +408,167 @@ export async function runWarpRouteApply(params: ApplyParams) {
...
+      return {
+        ...config,
+        ...tokenMetadata,
+      };
+    });
+
+    // Deploy and enrolls additional routers with each other
+    const newDeployedContracts = await executeDeploy(
+      {
+        // @TODO use EvmERC20WarpModule when it's ready
```
Comment (nbayindirli) [inline typescript/cli/src/deploy/warp.ts:479]: nit
```suggestion
        // TODO: use EvmERC20WarpModule when it's ready
```

File: typescript/cli/src/deploy/warp.ts [hunk 3]
```diff
@@ -387,48 +408,167 @@ export async function runWarpRouteApply(params: ApplyParams) {
...
+
+          if (transactions.length) {
+            for (const transaction of transactions) {
+              await multiProvider.sendTransaction(chain, transaction);
+            }
+
+            logGreen(`Warp config updated on ${chain}.`);
+          } else {
+            logGreen(
+              `Warp config on ${chain} is the same as target. No updates needed.`,
```
Comment (nbayindirli) [inline typescript/cli/src/deploy/warp.ts:442]: this is great

**Review by nbayindirli — COMMENTED** [review]
i thoroughly tested this PR, and although it is functional, I'm not sure we should consider this mergable until we add extend support to `warp init`. asking users to copy a new config into a previous config (or recreate the config from scratch with warp init) is error prone and actually pretty challenging since `configs/warp-route-deploy.yaml` is overwritten with each warp init

something like,
- run `warp init`
- "Pre-existing warp-route-deployment.yaml detected. Would you like to add to this config or create a new one?"
- if they'd like to extend, we append the newly created config to the previous

please let me know if there is another flow I'm missing to do what I described

additional comments on usability below

---

ran the following to test:
- `hl registry init: alpha`, success
- `hl registry init: beta`, success
- `hl core init`, success
- `hl core deploy: alpha`, success
- `hl core deploy: beta`, success
- `hl warp init: alpha (native) -> beta (synthetic)`, success
  - `warp-route-deployment.yaml` (current dir)
```
alpha:
  interchainSecurityModule:
    modules:
      - relayer: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: trustedRelayerIsm
      - domains: {}
        owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: "0xbE811734f5a88F3315e5Fae0619bbCAD2eD9a8B3"
  owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
  type: native
beta:
  interchainSecurityModule:
    modules:
      - relayer: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: trustedRelayerIsm
      - domains: {}
        owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: "0xB8cf7bb7f3298B293A9d58452ceB25CFCBED6f66"
  owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
  type: synthetic
```
- `hl warp deploy`, success
  - `alpha-beta-addresses.yaml` (local registry)
```
avalanche:
  native: "0x5F208a25410819c664fFDfFE4d8a10e1B1E0173B"
bsc:
  synthetic: "0x329FFF413F449dF665F651f31FC45F3f8D72155d"
```
  - `alpha-beta-config.yaml` (local registry)
```
# yaml-language-server: $schema=../schema.json
tokens:
  - addressOrDenom: "0x5F208a25410819c664fFDfFE4d8a10e1B1E0173B"
    chainName: avalanche
    connections:
      - token: ethereum|bsc|0x329FFF413F449dF665F651f31FC45F3f8D72155d
    decimals: 18
    name: Avalanche
    standard: EvmHypNative
    symbol: AVAX
  - addressOrDenom: "0x329FFF413F449dF665F651f31FC45F3f8D72155d"
    chainName: bsc
    connections:
      - token: ethereum|avalanche|0x5F208a25410819c664fFDfFE4d8a10e1B1E0173B
    decimals: 18
    name: Avalanche
    standard: EvmHypSynthetic
    symbol: AVAX
```
- [manually copied in a new alfajores (native) config that I created by hand, would have preferred running `hl warp init: alfajores (native)` on top of alpha (native) -> beta (synthetic)]
  - `warp-route-deployment.yaml` (current dir)
```
alpha:
  interchainSecurityModule:
    modules:
      - relayer: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: trustedRelayerIsm
      - domains: {}
        owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: "0xbE811734f5a88F3315e5Fae0619bbCAD2eD9a8B3"
  owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
  type: native
beta:
  interchainSecurityModule:
    modules:
      - relayer: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: trustedRelayerIsm
      - domains: {}
        owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: "0xB8cf7bb7f3298B293A9d58452ceB25CFCBED6f66"
  owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
  type: synthetic
alfajores:
  interchainSecurityModule:
    modules:
      - relayer: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: trustedRelayerIsm
      - domains: {}
        owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
        type: defaultFallbackRoutingIsm
    threshold: 1
    type: staticAggregationIsm
  isNft: false
  mailbox: "0xc4453203eD77C9848D38a4d623322ADaC5AbaB22"
  owner: "0x16F4898F47c085C41d7Cc6b1dc72B91EA617dcBb"
  type: native
```
_Calling out that at this point in time it is not intuitive if I should run hl warp deploy or apply, since even for deploy I expect I would be presented with a choice to create new or extend_

_It is also very odd that I must specify a warp route config (registry-side), when I have just made an update to the config in my current dir. Maybe that's just me_

- `hl warp apply`, success
  - `alfajores-alpha-beta-addresses.yaml` (local registry)
Screenshot 2024-07-18 at 4 48 32 PM
  - `alfajores-alpha-beta-config.yaml` (local registry)
Screenshot 2024-07-18 at 4 30 24 PM

_at this point, I am surprised to see the old `alpha-beta-config.yaml`, given that the new file is a superset. curious how others feel about this_

_also, if I were to run `hl warp deploy` right now, I would overwrite the new extend deployment. should we warn users "config already exists, overwrite?" when running warp deploy?_

---

super cool. my concerns are around usability and ux/safety.

cc @yorhodes @cmcewen @nambrot to continue the conversation

**Review by nbayindirli — APPROVED** [review]

File: typescript/cli/src/deploy/warp.ts (hunk 1 above)
Comment (ltyu) [reply typescript/cli/src/deploy/warp.ts]: fixed

File: typescript/cli/src/deploy/warp.ts (hunk 2 above)
Comment (ltyu) [reply typescript/cli/src/deploy/warp.ts]: fixed

File: typescript/cli/src/deploy/warp.ts [hunk 4]
```diff
@@ -185,12 +204,17 @@ async function executeDeploy(params: DeployParams, apiKeys: ChainMap<string>) {
 
   const deployedContracts = await deployer.deploy(modifiedConfig);
 
-  const warpCoreConfig = await getWarpCoreConfig(params, deployedContracts);
   logGreen('✅ Warp contract deployments complete');
+  return deployedContracts;
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/cli/src/deploy/warp.ts:212]: fixed

File: typescript/cli/src/deploy/warp.ts [hunk 5]
```diff
@@ -393,10 +423,11 @@ export async function runWarpRouteApply(params: ApplyParams): Promise<void> {
   WarpCoreConfigSchema.parse(warpCoreConfig);
   const addresses = await registry.getAddresses();
 
-  // Convert warpCoreConfig.tokens[] into a mapping of { [chainName]: Config }
-  // This allows O(1) reads within the loop
-  const warpCoreByChain = Object.fromEntries(
-    warpCoreConfig.tokens.map((token) => [token.chainName, token]),
+  const tokenConfigsByChain = Object.fromEntries(
```
Comment (ltyu) [inline typescript/cli/src/deploy/warp.ts:426]: not sure if i agree with "tokenConfig". arguably, there are 2 "tokenConfigs": WarpCore and WarpDeploy.

**Review by ltyu — COMMENTED** [review]
let me know next time before you make semantic changes so we are on the same page.

i like your choice of variables, which are way more explicit

File: typescript/cli/src/deploy/warp.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nbayindirli) [reply typescript/cli/src/deploy/warp.ts:426]: makes sense– can you update to `warpCoreTokenConfigsByChain`?

once again reminding me how unintuitive "warp core" vs "warp deploy" is. i cannot keep this straight

File: typescript/cli/src/deploy/warp.ts [hunk 6]
```diff
@@ -455,56 +486,59 @@ export async function runWarpRouteApply(params: ApplyParams): Promise<void> {
...
     );
 
-    // Derive additionalConfig metadata using the first deployed Warp Route.
-    const tokenMetadata = await HypERC20Deployer.deriveTokenMetadata(
+    const existingTokenMetadata = await HypERC20Deployer.deriveTokenMetadata(
       multiProvider,
-      deployedConfig,
+      existingConfig,
+    );
+    const mergedConfig: WarpRouteDeployConfig = objMap(
```
[the remark this answers is not in the document]
Comment (nbayindirli) [reply typescript/cli/src/deploy/warp.ts:503]: ah I see, cool let's revert

File: typescript/cli/src/deploy/warp.ts [hunk 7]
```diff
@@ -455,56 +486,59 @@ export async function runWarpRouteApply(params: ApplyParams): Promise<void> {
...
+      extendedConfig,
+      (_chain, extendedTokenMetadata) => {
```
[the remark this answers is not in the document]
Comment (nbayindirli) [reply typescript/cli/src/deploy/warp.ts:505]: why?

**Pull request merged**